File: apps/web/src/logic/heading.ts
import { sanitizeTelemetrySample, type TelemetrySample } from './telemetry'

export const UNKNOWN_GLOBAL_HEADING_VALUE = 65535

export type HeadingSource = 'VFR_HUD.heading' | 'ATTITUDE.yaw' | 'GLOBAL_POSITION_INT.hdg' | 'none'

export interface HeadingResolution {
  headingDeg: number | null
  source: HeadingSource
  isFallback: boolean
}

export function normalizeHeadingDegrees(input: number): number {
  const normalized = input % 360
  return normalized < 0 ? normalized + 360 : normalized
}

export function yawRadiansToHeadingDegrees(yawRad: number): number {
  const yawDeg = (yawRad * 180) / Math.PI
  return normalizeHeadingDegrees(yawDeg)
}

function readVfrHeading(sample: TelemetrySample): number | undefined {
  return sample.vfrHud?.headingDeg
}

function readAttitudeYawHeading(sample: TelemetrySample): number | undefined {
  const yawRad = sample.attitude?.yawRad

  if (yawRad === undefined) {
    return undefined
  }

  return yawRadiansToHeadingDegrees(yawRad)
}

function readGlobalPositionHeading(sample: TelemetrySample): number | undefined {
  const headingCdeg = sample.globalPositionInt?.headingCdeg

  if (headingCdeg === undefined || headingCdeg === UNKNOWN_GLOBAL_HEADING_VALUE) {
    return undefined
  }

  return normalizeHeadingDegrees(headingCdeg / 100)
}

export function resolveHeading(sampleInput: TelemetrySample): HeadingResolution {
  const sample = sanitizeTelemetrySample(sampleInput)

  const vfrHeading = readVfrHeading(sample)
  if (vfrHeading !== undefined) {
    return {
      headingDeg: normalizeHeadingDegrees(vfrHeading),
      source: 'VFR_HUD.heading',
      isFallback: false,
    }
  }

  const attitudeHeading = readAttitudeYawHeading(sample)
  if (attitudeHeading !== undefined) {
    return {
      headingDeg: attitudeHeading,
      source: 'ATTITUDE.yaw',
      isFallback: true,
    }
  }

  const globalHeading = readGlobalPositionHeading(sample)
  if (globalHeading !== undefined) {
    return {
      headingDeg: globalHeading,
      source: 'GLOBAL_POSITION_INT.hdg',
      isFallback: true,
    }
  }

  return {
    headingDeg: null,
    source: 'none',
    isFallback: false,
  }
}

File: apps/esp32/lib/hud/src/heading.cpp
#include "hud/heading.hpp"

#include <cmath>

namespace hud {
namespace {
constexpr float kRadiansToDegrees = 57.2957795131F;

bool finite(float value) {
  return std::isfinite(value);
}
}  // namespace

float normalize_degrees(float degrees) {
  const float normalized = std::fmod(degrees, 360.0F);
  return normalized < 0.0F ? normalized + 360.0F : normalized;
}

HeadingResolution resolve_heading(const TelemetrySample& sample) {
  if (sample.vfr_hud.has_value() && sample.vfr_hud->heading_deg.has_value() &&
      finite(*sample.vfr_hud->heading_deg)) {
    return {normalize_degrees(*sample.vfr_hud->heading_deg), HeadingSource::kVfrHud, false};
  }

  if (sample.attitude.has_value() && sample.attitude->yaw_rad.has_value() &&
      finite(*sample.attitude->yaw_rad)) {
    return {normalize_degrees(*sample.attitude->yaw_rad * kRadiansToDegrees),
            HeadingSource::kAttitudeYaw, true};
  }

  if (sample.global_position.has_value() && sample.global_position->heading_cdeg.has_value() &&
      *sample.global_position->heading_cdeg != 65535U) {
    return {normalize_degrees(static_cast<float>(*sample.global_position->heading_cdeg) / 100.0F),
            HeadingSource::kGlobalPosition, true};
  }

  return {std::nullopt, HeadingSource::kUnavailable, false};
}

const char* heading_source_name(HeadingSource source) {
  switch (source) {
    case HeadingSource::kVfrHud:
      return "VFR_HUD.heading";
    case HeadingSource::kAttitudeYaw:
      return "ATTITUDE.yaw";
    case HeadingSource::kGlobalPosition:
      return "GLOBAL_POSITION_INT.hdg";
    case HeadingSource::kUnavailable:
      return "unavailable";
  }
  return "unavailable";
}

}  // namespace hud
